plugin/pkg/admission/webhook/admission_test.go

```diff
@@ -0,0 +1,363 @@
+					},
+				},
+			},
+			value: nil,
+		},
+		GenericAdmissionWebhookTest{
+			test:   "Matching rule sends (webhook internal server error)",
+			config: makeGoodConfig(),
+			value:  fmt.Errorf(`%san error on the server ("webhook internal server error") has prevented the request from succeeding`, errAdmitPrefix),
+		},
+		GenericAdmissionWebhookTest{
+			test:   "Matching rule sends (webhook unsuccessful response)",
+			config: makeGoodConfig(),
+			value:  fmt.Errorf("%serror contacting webhook: 101", errAdmitPrefix),
+		},
+		GenericAdmissionWebhookTest{
+			test:   "Matching rule sends (webhook unmarshallable response)",
+			config: makeGoodConfig(),
+			value:  fmt.Errorf("%scouldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct.*", errAdmitPrefix),
+		},
+		GenericAdmissionWebhookTest{
+			test:   "Matching rule sends (webhook error but allowed via fail action)",
+			config: configWithFailAction,
+			value:  nil,
+		},
+		GenericAdmissionWebhookTest{
+			test:   "Matching rule sends (webhook request denied without reason)",
+			config: makeGoodConfig(),
+			value:  fmt.Errorf("%swebhook backend denied the request", errAdmitPrefix),
+		},
+		GenericAdmissionWebhookTest{
+			test:   "Matching rule sends (webhook request denied with reason)",
+			config: makeGoodConfig(),
+			value:  fmt.Errorf("%swebhook backend denied the request: you shall not pass", errAdmitPrefix),
+		},
+		GenericAdmissionWebhookTest{
+			test:   "Matching rule sends (webhook request allowed)",
+			config: makeGoodConfig(),
+			value:  nil,
+		},
+	}
+
+	for _, tt := range tests {
+		wh, err := NewGenericAdmissionWebhook(tt.config)
+
+		if err != nil {
+			t.Errorf("%s: unexpected error: %v", tt.test, err)
+		} else {
+			err = wh.Admit(admission.NewAttributesRecord(&object, &oldObject, kind, namespace, name, resource, subResource, operation, &userInfo))
+
+			checkForError(t, tt.test, tt.value, err)
+		}
+	}
+}
+
+func checkForError(t *testing.T, test string, expected, actual interface{}) {
+	aErr, _ := actual.(error)
+	eErr, _ := expected.(error)
+
+	if eErr != nil {
+		if aErr == nil {
+			t.Errorf("%s: expected an error", test)
+		} else if eErr.Error() != aErr.Error() && !regexp.MustCompile(eErr.Error()).MatchString(aErr.Error()) {
+			t.Errorf("%s: unexpected error message to match:\n  Expected: %s\n  Actual:   %s", test, eErr.Error(), aErr.Error())
+		}
+	} else {
+		if aErr != nil {
+			t.Errorf("%s: unexpected error: %v", test, aErr)
+		}
+	}
+}
+
+func killTests(err error) {
+	panic(fmt.Sprintf("Unable to bootstrap tests: %v", err))
+}
+
+func makeGoodConfig() *GenericAdmissionWebhookConfig {
+	return &GenericAdmissionWebhookConfig{
+		KubeConfigFile: filepath.Join(testRoot, "kube-config"),
+		Rules: []Rule{
+			Rule{
+				Operations: []admission.Operation{
+					admission.Update,
+				},
+				Type: Send,
+			},
+		},
+	}
+}
+
+func webhookHandler(w http.ResponseWriter, r *http.Request) {
+	requestCount++
```

Comment:
Well, since the tests are synchronous it seemed simple.  Last thing I wanted to do was spin up a test server for each test, which would also require the extra boilerplate for defining the test-specific handler in the test table.